java/ql/src/semmle/code/java/dataflow/FlowSteps.qll

```diff
@@ -0,0 +1,71 @@
+/**
+ * Provides classes representing various flow steps for taint tracking.
+ */
+
+private import java
+private import semmle.code.java.dataflow.DataFlow
+
+/**
+ * A module importing the frameworks that implement additional flow steps,
+ * ensuring that they are visible to the taint tracking library.
+ */
+module Frameworks {
+  private import semmle.code.java.frameworks.jackson.JacksonSerializability
+  private import semmle.code.java.frameworks.android.Intent
+  private import semmle.code.java.frameworks.android.SQLite
+  private import semmle.code.java.frameworks.Guice
+  private import semmle.code.java.frameworks.Protobuf
+}
+
+/**
+ * A unit class for adding additional taint steps.
+ *
+ * Extend this class to add additional taint steps that should apply to all
+ * taint configurations.
+ */
+class AdditionalTaintStep extends Unit {
+  /**
+   * Holds if the step from `node1` to `node2` should be considered a taint
+   * step for all configurations.
+   */
+  abstract predicate step(DataFlow::Node node1, DataFlow::Node node2);
+}
+
+/**
+ * A method that preserves taint.
+ *
+ * Extend this class and override at least one of `returnsTaintFrom` or `transfersTaint`
+ * to add additional taint steps through a method that should apply to all taint configurations.
+ */
+abstract class TaintPreservingMethod extends Method {
+  /**
+   * Holds if this method returns tainted data when `arg` tainted.
+   * `arg` is a parameter index, or is -1 to indicate the qualifier.
+   */
+  predicate returnsTaintFrom(int arg) { none() }
+
+  /**
+   * Holds if this method writes tainted data to `sink` when `src` is tainted.
+   * `src` and `sink` are parameter indices, or -1 to indicate the qualifier.
+   */
+  predicate transfersTaint(int src, int sink) { none() }
+}
+
+private class StringTaintPreservingMethod extends TaintPreservingMethod {
+  StringTaintPreservingMethod() {
+    this.getDeclaringType() instanceof TypeString and
+    this
+        .hasName(["concat", "copyValueOf", "endsWith", "format", "formatted", "getBytes", "indent",
+              "intern", "join", "repeat", "split", "strip", "stripIndent", "stripLeading",
+              "stripTrailing", "substring", "toCharArray", "toLowerCase", "toString", "toUpperCase",
+              "trim"])
+  }
+
+  override predicate returnsTaintFrom(int arg) {
+    arg = -1
```

Comment:
```suggestion
    arg = -1 and not this.isStatic()
```
I prefer this addition. It's mostly for clarity, but `"tainted".copyValeOf(x)` is valid Java (even though it's a bit obscure).

Reply:
Good point, I didn't know that was valid.